### PR TITLE
[flink]Supported ogg format and optimize format scalability

### PR DESCRIPTION
**Pull request by MonsterChenzhuo**
### Purpose
subtask issue: #1826

This PR accomplishes the following:
It completes the decoupling work of the format. Through the abstraction of DataFormat, integrating a new format has been simplified.

For instance, if you wish to develop a new format, you only need to follow two steps, without altering or modifying the existing code logic:

Implement the parsing logic for the new format. You can refer to my implementation at: org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java.
Add this new format to DataFormat.

### Tests
IT：KafkaOggSyncDatabaseActionITCase
图片

IT：KafkaOggSyncTableActionITCase
图片

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Very thanks to @MonsterChenzhuo .

This is what we want to do! We must do some abstraction for log sync.
1. Abstraction for format: canal debezium and more...
2. Abstraction for log system: kafka pulsar..

This is the first step.

**Review by JingsongLi — COMMENTED** [review]
Can you try to implement a parser for debezium?
We should take a look if this refactor is OK for other formats.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java:60]: I think this interface can be named `RecordParserFactory`.

**Comment by MonsterChenzhuo** [comment]
> Can you try to implement a parser for debezium? We should take a look if this refactor is OK for other formats.

Sure, I can do that.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java]: done.

**Comment by MonsterChenzhuo** [comment]
> Very thanks to @MonsterChenzhuo .
> 
> This is what we want to do! We must do some abstraction for log sync.
> 
>     1. Abstraction for format: canal debezium and more...
> 
>     2. Abstraction for log system: kafka pulsar..
> 
> 
> This is the first step.
@JingsongLi @yuzelin  PTAL,thanks.

**Comment by Alibaba-HZY** [comment]
very well job,ogg-json does not contain field type information, all types are string, whether this does not meet user expectation? but there doesn't seem to be a better solution

**Comment by MonsterChenzhuo** [comment]
> very well job,ogg-json does not contain field type information, all types are string, whether this does not meet user expectation? but there doesn't seem to be a better solution

In my personal experience, string types work for most situations. However, I have a follow up plan for your concern.
I plan to implement it later:
1. In the whole library synchronization scenario, allow users to create tables in Paimon first. Synchronization will be based on the table created by the user for type mapping.
2. Table level synchronization. This will allow users to pass data type parameters. This approach is similar to FlinkSQL.

**Comment by MonsterChenzhuo** [comment]
> Can you try to implement a parser for debezium? We should take a look if this refactor is OK for other formats.



> Can you try to implement a parser for debezium? We should take a look if this refactor is OK for other formats.

Currently, the pr has chosen ogg for its implementation. After the merge, I will continue to push for debezium support.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:214]: Computed column value is evaluated, should pad the paimonFieldTypes?
I also noticed that CanalRecordParser `extractRecords` has the same problem, please fix it.

**Review by yuzelin — COMMENTED** [review]
Thanks for contributing! I left some comments.

File: docs/content/how-to/cdc-ingestion.md
Comment (MonsterChenzhuo) [reply docs/content/how-to/cdc-ingestion.md:305]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java:126]: done.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaActionITCaseBase.java
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaActionITCaseBase.java:388]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:214]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:76]: done.

**Comment by MonsterChenzhuo** [comment]
@JingsongLi @yuzelin  PTAL,thanks.

**Comment by yuzelin** [comment]
LGTM. cc @JingsongLi

**Comment by MonsterChenzhuo** [comment]
> LGTM. cc @JingsongLi
@yuzelin 
The problem of ci is a bit of a headache。I want to migrate the content of CDC into the paimon-flink/paimon-flink-cdc-e2e module in order to address the timeout issue of the utitcase-flink.yml CI process. Do you have any suggestions?

**Comment by yuzelin** [comment]
How about use syntax `mvn -T 1C -B clean install -pl 'org.apache.paimon:paimon-flink-common' -Dtest='!org.apache.paimon.flink.action.*'` in flink tests? I think this can skip the action tests. Then we can use `mvn test -Dxxx` to specify the action tests. WDYT?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**
